### PR TITLE
Document compact ranges application in a witness

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
Part of #4.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
https://github.com/transparency-dev/merkle/blob/witness/docs/compact_ranges.md

File: docs/compact_ranges.md [hunk 1]
```diff
@@ -210,3 +210,37 @@ the load on the coordinator with the load on "leaf" workers.
 
 The described approach can be used to implement a scalable tamper-evident log
 based on Merkle trees.
+
+### Updatable Proofs
+
+Logs based on Merkle trees are rarely static - usually they grow in append-only
+fashion. Proofs built for one state of the tree get outdated with later states.
+If proofs are made using compact ranges, they can be updated by merging in the
```
Comment (AlCutter) [inline docs/compact_ranges.md:218]: "When proofs are represented using compact ranges..."?

File: docs/compact_ranges.md [hunk 2]
```diff
@@ -210,3 +210,37 @@ the load on the coordinator with the load on "leaf" workers.
...
+compact range between the previous and the next tree size.
+
+For example, an inclusion proof for range `[L, R)` in a tree of size `N` can be
+represented with compact ranges `[0, L)` and `[R, N)`, as described in a
```
Comment (AlCutter) [inline docs/compact_ranges.md:222]: /in a [section]/in the section on [arbitrary inclusion proofs] above/"

File: docs/compact_ranges.md [hunk 3]
```diff
@@ -210,3 +210,37 @@ the load on the coordinator with the load on "leaf" workers.
...
+[section](#arbitrary-inclusion-proofs) above. When the tree moves to a state of
+size `N+delta`, it is possible to update this inclusion proof by merging the
+compact range `[R, N)` with `[N, N+delta)`.
+
+Updating a proof incrementally is more efficient in terms of space / bandwidth
+than fetching an entire proof each time. For example, a consistency proof in a
+tree up to size `N` takes `O(log N)` space, and fetching it `O(N)` time takes
```
Comment (AlCutter) [inline docs/compact_ranges.md:229]: /tree up to size N/tree of size N/?

Comment (AlCutter) [inline docs/compact_ranges.md:229]: O(N) time**s**?

File: docs/compact_ranges.md [hunk 4]
```diff
@@ -210,3 +210,37 @@ the load on the coordinator with the load on "leaf" workers.
...
+`O(N log N)`. Incremental updates with compact ranges will total up as `O(N)`
+since each of the tree nodes is used in at most one compact range in the series.
+
+### Witness
+
+A witness is an entity that watches the state of the log and attests to its
+consistency. As described in a [section](#root-hash) above, the tree state can
```
Comment (AlCutter) [inline docs/compact_ranges.md:236]: "in the [Root Hash section] above"

**Review by AlCutter — APPROVED** [review]
Few nits inline

File: docs/compact_ranges.md (hunk 1 above)
Comment (pavelkalinnikov) [reply docs/compact_ranges.md]: Done.

File: docs/compact_ranges.md (hunk 2 above)
Comment (pavelkalinnikov) [reply docs/compact_ranges.md]: Done.

File: docs/compact_ranges.md (hunk 3 above)
Comment (pavelkalinnikov) [reply docs/compact_ranges.md]: The idea is that the tree grows gradually to size `N`, I was avoiding fixing the size.

Reworded a bit, how does it look?

Comment (pavelkalinnikov) [reply docs/compact_ranges.md]: Done.

File: docs/compact_ranges.md (hunk 4 above)
Comment (pavelkalinnikov) [reply docs/compact_ranges.md]: Done.

**Review by pavelkalinnikov — COMMENTED** [review]
Addressed Al's comments.

File: docs/compact_ranges.md (hunk 3 above)
Comment (AlCutter) [reply docs/compact_ranges.md]: Yeah, that's clearer - ta!

**Pull request merged**